### PR TITLE
feat(insights): implement routing for frontend domain view

### PR DESCRIPTION
**Pull request by DominikB2014**
Work for https://github.com/getsentry/sentry/issues/77572

This PR ensures that all module links within the frontend domain view, stay within the domain view.
1. Creates a shared frontend header which contains breadcrumbs and tabs
2. Create a `useDomainFilters` hook which returns the current view and whether the current route is within the domain view
3. Conditionally render shared frontend header on frontend modules based on if its in the domain view
4. Update `useModuleUrl` to support new domain view urls
5. Add routes for domain view modules

### CODE REVIEW

**Review by 0Calories — CHANGES_REQUESTED** [review]
I tested the frontend domain locally and noticed some bugs:

- Switching tabs in the Frontend domain keeps appending route segments, so you end up with URLs like this: `https://sentry.sentry.io/performance/frontend/performance/frontend/browser/performance/frontend/browser/pageloads/performance/frontend/browser/pageloads/performance/frontend/performance/frontend/http/performance/frontend/http/performance/frontend/browser/performance/frontend/browser/pageloads/performance/frontend/browser/pageloads/performance/frontend/performance/frontend/http/performance/frontend/http/?project=11276&project=1473210&statsPeriod=14d`
- Switching tabs causes the entire page to refresh instead of just re-rendering the body

I did mention before I don't have a strong preference on using a route path segment vs a query param for the tab changes, but I forgot to consider that the former causes the entire page to refresh. Maybe I'm missing some context, but looking at the code it doesn't seem like it would be hard to stick to using a query param to determine which module to display for the current domain. You just would need to use `navigate`, copying the current URL and adding a param instead of `moduleURLBuilder`

File: static/app/views/insights/pages/frontend/frontendPageHeader.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    if (key === OVERVIEW_PAGE_TITLE) {
+      navigate(frontendBaseUrl);
+      return;
+    }
+    navigate(moduleURLBuilder(key as RoutableModuleNames));
+  };
+
+  return (
+    <Fragment>
+      <Tabs value={module || OVERVIEW_PAGE_TITLE} onChange={handleTabChange}>
```
Comment (0Calories) [inline static/app/views/insights/pages/frontend/frontendPageHeader.tsx:73]:
```suggestion
      <Tabs value={module ?? OVERVIEW_PAGE_TITLE} onChange={handleTabChange}>
```
nit: I like the nullish coalescing operator for these cases to avoid the potential weird case of these expressions returning a boolean

File: static/app/routes.tsx [hunk 2]
```diff
@@ -1641,10 +1641,61 @@ function buildRoutes() {
         path="trends/"
         component={make(() => import('sentry/views/performance/trends'))}
       />
-      <Route
-        path={`${FRONTEND_LANDING_SUB_PATH}/`}
-        component={make(() => import('sentry/views/insights/pages/frontendLandingPage'))}
-      />
+      <Route path={`${FRONTEND_LANDING_SUB_PATH}/`}>
```
[the remark this answers is not in the document]
Comment (DominikB2014) [reply static/app/routes.tsx:1644]: It should be fine to import this. It all comes for a settings.ts file which only imports the translate function. The translate function is already imported into the routes.tsx file, so its not like we're bringing in anything extra into routes bundle. 

The purpose of this constant is for this exact reason, we've gone through a bunch of renaming urls and links in the past, and it was a pain everytime, stuff like this make should make it a little easier.

File: static/app/views/insights/pages/frontend/frontendPageHeader.tsx (hunk 1 above)
Comment (DominikB2014) [reply static/app/views/insights/pages/frontend/frontendPageHeader.tsx]: Fair! But i'm curious when this would happen, `module` is of type `ModuleName`, wouldn't typescript prevent these weird scenarios?

Comment (0Calories) [reply static/app/views/insights/pages/frontend/frontendPageHeader.tsx]: It wouldn't happen in this particular case but I'm averse to using it unless we have to, and have been warned about it in the past. The or operator is perfectly fine here, it just makes me uncomfortable hence the nitpick lol

File: static/app/routes.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (0Calories) [reply static/app/routes.tsx:1644]: Fair enough!

**Comment by DominikB2014** [comment]
@0Calories
> * Switching tabs in the Frontend domain keeps appending route segments, so you end up with URLs like this:

Fixed this!

> * Switching tabs causes the entire page to refresh instead of just re-rendering the body
Looking into this
 
> I did mention before I don't have a strong preference on using a route path segment vs a query param for the tab changes, but I forgot to consider that the former causes the entire page to refresh. Maybe I'm missing some context, but looking at the code it doesn't seem like it would be hard to stick to using a query param to determine which module to display for the current domain. You just would need to use `navigate`, copying the current URL and adding a param instead of `moduleURLBuilder`
You're right that you could definetly update the header to use navigate instead of moduleURLBuilder, but the header is not the issue. The issue is there is existing links all around the app that rely on using `useModuleURL` to get to the module's landing page and subpages. For example
https://github.com/getsentry/sentry/blob/257ac7ad71cda1ebb95cbcb9bdce830eac95b11f/static/app/views/insights/http/components/tables/domainCell.tsx#L30
`moduleURL` here would be `/insights/http` giving a final url of `/insights/http/domains/?....`
For the the domain view we would want something like `/performance/http/domains?...` so moduleURL could just return `/performance/http`

However in a case like this
https://github.com/getsentry/sentry/blob/257ac7ad71cda1ebb95cbcb9bdce830eac95b11f/static/app/components/events/interfaces/llm-monitoring/llmMonitoringSection.tsx#L83
were linking to `moduleURL` directly, which would be `/insights/ai`
If we went with query params, we would have to return `/performance/ai?module=llm`

So basically for subpages we would have `/performance/<view>/<subpage>/
and for landingpage we would have `/performance/<view>?module=<module>`

As `useModuleURL` isn't intended to know if you want a landing page or subpage, we'll have to go through the entire app, or perhaps setup redirects to fix this

**Comment by 0Calories** [comment]
Ah that makes a lot more sense now, thanks for clarifying. I get why having to use queryParams would be a painful experience now due to our reliance of `useModuleURL` in all of Insights. I still don't like how the entire page re-renders with tab changes but I understand the pain behind fixing something like that. We can consider it later down the line I suppose.

One more thing: The backend Outbound API requests tab has double the headers
image

**Comment by DominikB2014** [comment]
@0Calories I believe I address all your concerns!

**Comment by 0Calories** [comment]
Yeah, the flickering is happening because react router is recognizing each tab as a new page altogether since they have separate routes. It only flickers on the first visit because the changes are cached. This isn't the case in most parts of the product, namely in places where tabs are treated as a change in the query param rather than a new route (see current performance landing page, Discover, Profiles).

It's fine though, let's think about it a bit further down the line, I may also take a deeper look into how much work it would take to refactor `useModuleURL` to consider modules as a query param while not breaking the links in other places in Insights where it is used.

File: static/app/views/insights/pages/useFilters.tsx [hunk 3]
```diff
@@ -3,16 +3,46 @@ import type {ModuleName} from 'webpack-cli';
...
+  view?: DomainView;
+};
 
 export type Filters = {
   module?: ModuleName;
 };
 
+export const useDomainViewFilters = () => {
+  const location = useLocation();
+  const pathSegements = location.pathname.split('/').filter(Boolean);
```
Comment (0Calories) [inline static/app/views/insights/pages/useFilters.tsx:28]:
```suggestion
  const pathSegments = location.pathname.split('/').filter(Boolean);
```

File: static/app/views/insights/pages/useFilters.tsx [hunk 4]
```diff
@@ -3,16 +3,46 @@ import type {ModuleName} from 'webpack-cli';
...
+  const indexOfPerformance = pathSegements.indexOf('performance');
```
Comment (0Calories) [inline static/app/views/insights/pages/useFilters.tsx:29]:
```suggestion
  const indexOfPerformance = pathSegments.indexOf('performance');
```

File: static/app/views/insights/pages/useFilters.tsx [hunk 5]
```diff
@@ -3,16 +3,46 @@ import type {ModuleName} from 'webpack-cli';
...
+  const isInDomainView = indexOfPerformance !== -1;
+
+  const view = pathSegements[indexOfPerformance + 1] as DomainViewFilters['view'];
```
Comment (0Calories) [inline static/app/views/insights/pages/useFilters.tsx:32]:
```suggestion
  const view = pathSegments[indexOfPerformance + 1] as DomainViewFilters['view'];
```

**Review by 0Calories — APPROVED** [review]
